File: app/services/azure.js
import axios from 'axios';
import FormData from 'form-data';

/**
 * Azure OpenAI REST helpers
 *
 * Endpoints:
 * - Chat Completions: /openai/deployments/{deployment}/chat/completions?api-version=...
 * - Transcriptions:   /openai/deployments/{deployment}/audio/transcriptions?api-version=...
 * - Text-to-Speech:   /openai/deployments/{deployment}/audio/speech?api-version=...
 *
 * References:
 * - Whisper transcription endpoint pattern shown in Microsoft Learn quickstart. 
 * - TTS /audio/speech endpoint and request body shown in Microsoft Learn quickstart.
 */

function requireEnv(name) {
  const v = process.env[name];
  if (!v) throw new Error(`Missing required env var: ${name}`);
  return v;
}

function baseUrl() {
  // Ensure no trailing slash issues
  const endpoint = requireEnv('AZURE_OPENAI_ENDPOINT').replace(/\/+$/, '');
  return endpoint;
}

function apiHeaders(extra = {}) {
  return {
    'api-key': requireEnv('AZURE_OPENAI_API_KEY'),
    ...extra
  };
}

function apiVersion() {
  return process.env.AZURE_OPENAI_API_VERSION || '2025-04-01-preview';
}

export async function generateText(userPrompt) {
  const deployment = requireEnv('AZURE_GPT_DEPLOYMENT');

  const url = `${baseUrl()}/openai/deployments/${encodeURIComponent(deployment)}/chat/completions?api-version=${encodeURIComponent(apiVersion())}`;

  const body = {
    messages: [
      { role: 'system', content: 'You are a helpful WhatsApp assistant. Keep replies concise.' },
      { role: 'user', content: String(userPrompt ?? '') }
    ],
    temperature: 0.3
  };

  const r = await axios.post(url, body, {
    headers: { ...apiHeaders(), 'Content-Type': 'application/json' },
    timeout: 120000
  });

  const text = r.data?.choices?.[0]?.message?.content;
  return (text || '').trim() || 'Sorry, I could not generate a response.';
}

export async function transcribeAudio(audioBuffer) {
  const deployment = requireEnv('AZURE_STT_DEPLOYMENT');

  const url = `${baseUrl()}/openai/deployments/${encodeURIComponent(deployment)}/audio/transcriptions?api-version=${encodeURIComponent(apiVersion())}`;

  const form = new FormData();
  // WhatsApp audio is often OGG/Opus; Azure accepts common audio containers.
  form.append('file', audioBuffer, { filename: 'in.ogg', contentType: 'audio/ogg' });

  const r = await axios.post(url, form, {
    headers: { ...apiHeaders(), ...form.getHeaders() },
    maxBodyLength: Infinity,
    timeout: 300000
  });

  const text = r.data?.text;
  return (text || '').trim() || '';
}

export async function synthesizeSpeech(text) {
  const deployment = requireEnv('AZURE_TTS_DEPLOYMENT');

  const url = `${baseUrl()}/openai/deployments/${encodeURIComponent(deployment)}/audio/speech?api-version=${encodeURIComponent(apiVersion())}`;

  const voice = process.env.AZURE_TTS_VOICE || 'alloy';
  const format = process.env.AZURE_TTS_FORMAT || 'mp3';

  // Per Microsoft Learn, request body includes model, input, voice (and can include format).
  const body = {
    model: deployment, // Azure SDK examples pass deploymentName as model.
    input: String(text ?? ''),
    voice,
    format
  };

  const r = await axios.post(url, body, {
    headers: { ...apiHeaders(), 'Content-Type': 'application/json' },
    responseType: 'arraybuffer',
    timeout: 300000
  });

  return Buffer.from(r.data);
}

File: api/azure.py
import os
import mimetypes

import httpx
from fastapi import HTTPException


def require_env(name: str) -> str:
  value = os.getenv(name)
  if not value:
    raise RuntimeError(f"Missing required env var: {name}")
  return value


def base_url() -> str:
  return require_env("AZURE_OPENAI_ENDPOINT").rstrip("/")


def api_version() -> str:
  return os.getenv("AZURE_OPENAI_API_VERSION", "2025-04-01-preview")


def api_headers() -> dict:
  return {"api-key": require_env("AZURE_OPENAI_API_KEY")}


def audio_content_type() -> str:
  fmt = os.getenv("AZURE_TTS_FORMAT", "mp3").lower()
  is_mp3 = fmt in ("mp3", "mpeg", "audio/mpeg")
  return "audio/mpeg" if is_mp3 else "audio/ogg"


async def generate_text(user_prompt: str) -> str:
  deployment = require_env("AZURE_GPT_DEPLOYMENT")
  url = f"{base_url()}/openai/deployments/{deployment}/chat/completions"
  params = {"api-version": api_version()}
  body = {
    "messages": [
      {"role": "system", "content": "You are a helpful assistant. Keep replies concise."},
      {"role": "user", "content": str(user_prompt or "")},
    ],
    "temperature": 0.3,
  }

  async with httpx.AsyncClient(timeout=120) as client:
    r = await client.post(url, params=params, headers=api_headers(), json=body)
    try:
      r.raise_for_status()
    except httpx.HTTPStatusError as exc:
      detail = exc.response.text
      raise HTTPException(status_code=502, detail=f"Azure GPT error: {detail}") from exc
    text = r.json().get("choices", [{}])[0].get("message", {}).get("content", "")
    text = (text or "").strip()
    return text or "Sorry, I could not generate a response."


async def transcribe_audio(audio_bytes: bytes, filename: str, content_type: str | None) -> str:
  deployment = require_env("AZURE_STT_DEPLOYMENT")
  url = f"{base_url()}/openai/deployments/{deployment}/audio/transcriptions"
  params = {"api-version": api_version()}
  inferred = content_type
  if not inferred:
    inferred, _ = mimetypes.guess_type(filename or "")
  files = {"file": (filename or "audio", audio_bytes, inferred or "application/octet-stream")}

  async with httpx.AsyncClient(timeout=300) as client:
    r = await client.post(url, params=params, headers=api_headers(), files=files)
    try:
      r.raise_for_status()
    except httpx.HTTPStatusError as exc:
      detail = exc.response.text
      raise HTTPException(status_code=502, detail=f"Azure STT error: {detail}") from exc
    text = r.json().get("text", "")
    return (text or "").strip()


async def synthesize_speech(text: str) -> bytes:
  deployment = require_env("AZURE_TTS_DEPLOYMENT")
  url = f"{base_url()}/openai/deployments/{deployment}/audio/speech"
  params = {"api-version": api_version()}
  body = {
    "model": deployment,
    "input": str(text or ""),
    "voice": os.getenv("AZURE_TTS_VOICE", "alloy"),
    "format": os.getenv("AZURE_TTS_FORMAT", "mp3"),
  }

  async with httpx.AsyncClient(timeout=300) as client:
    r = await client.post(url, params=params, headers=api_headers(), json=body)
    try:
      r.raise_for_status()
    except httpx.HTTPStatusError as exc:
      detail = exc.response.text
      raise HTTPException(status_code=502, detail=f"Azure TTS error: {detail}") from exc
    return r.content
